websocket: WebSocketHandler.on_message allows coroutines

ajdavis opened the pull request:
websocket: WebSocketHandler.on_message allows coroutines

Fixes #1650

bdarnell commented:
This will *allow* coroutines for `on_message`, but by spawning a new callback for each one, those coroutines could run "concurrently" with no backpressure. I think it would be better (and more consistent with streaming callbacks in other parts of Tornado) if new messages couldn't be handled until the previous coroutine had returned (users could always opt in to calling spawn_callback themselves if they want that behavior). 

The simplest way to do this would be via plumbing in `_handle_message`, so we don't transition to the next `_receive_frame` state until the `on_message` callback has completed. That would block handling of control frames (ping, pong, close) while the callback is running. I think that's probably fine, although if we wanted to avoid it we could put the callback on a queue and not block the frame processing until the queue gets full.

ajdavis commented on tornado/websocket.py:
```diff
@@ -815,12 +828,18 @@ def _on_frame_data(self, data):
+            if handled_future:
+                # on_message is a coroutine, process more frames once it's done.
+                gen.convert_yielded(handled_future).add_done_callback(
+                    lambda future: self._receive_frame())
```
This is the significant change.

ajdavis commented on tornado/test/websocket_test.py:
```diff
@@ -260,6 +270,16 @@ def test_path_args(self):
+        self.assertEqual(res, 'hello1')
+        yield ws.write_message('hello2')
+        res = yield ws.read_message()
+        self.assertEqual(res, 'hello2')
```
This test, and test_native_coroutine below, now write two messages and read two responses, instead of one.

ajdavis commented:
Done, and I hope I got it right. How should the new tests prove that the server is waiting for the on_message coroutine to finish? If I could override the write buffer size of the tests' WebSocketClientConnections, then maybe I could assert that the Future returned by `ws.write_message` in the tests isn't marked "done" until the on_message coroutine finishes?

jeyraof commented:
When this feature will be deployed?

ajdavis commented:
Hey Ben, I think this can go in, but take a look at my previous comment about testing.

bdarnell commented on tornado/test/websocket_test.py:
```diff
@@ -92,6 +93,13 @@ def open(self, arg):
+class CoroutineOnMessageHandler(TestWebSocketHandler):
+    @gen.coroutine
+    def on_message(self, message):
+        yield gen.sleep(0.1)
```
I think what I'd do for testing is add a check that only one coroutine is sleeping at a time:

    if self.sleeping > 0:
        self.write_message('another coroutine is already sleeping')
    self.sleeping += 1
    yield gen.sleep(0.01)
    self.sleeping -= 1

And then in the tests themselves, write both messages before reading the first response.

ajdavis commented:
Ben I've implemented your testing idea, squashed and rebased on master (handled a merge conflict with https://github.com/tornadoweb/tornado/commit/2a157f41480f29131983758a61fb988c33443399).
